### PR TITLE
feat: formal verification & fuzz campaign

### PR DESCRIPTION
**Pull request by drgorillamd**

### CODE REVIEW

File: .github/workflows/coverage_check.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        with:
+          node-version: 18.x
+          cache: 'yarn'
+
+      - name: Install dependencies
+        run: yarn --frozen-lockfile --network-concurrency 1
+
+      - name: Run coverage
+        shell: bash
+        run: forge coverage --report summary --report lcov
```
[the remark this answers is not in the document]
Comment (gas1cent) [reply .github/workflows/coverage_check.yml:31]: Is `--report summary --report lcov` a quirk? The different order of arguments does not produce the same result.

[the remark this answers is not in the document]
Comment (drgorillamd) [reply .github/workflows/coverage_check.yml:31]: I'm filtering test and script directory with lcov, I couldn't manage to exclude them with a regex...

summary then lcov will display the summary table and create the lcov file, while the other way around doesn't display the summary (but, for me, creates the same lcov file tho). You have other differences?

File: .github/workflows/coverage_check.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            echo "Coverage decreased from $old to $new"
+            exit 1
+          fi
+          mv lcovNew.info coverage.info
+
+      - name: Upload the new coverage
+        uses: actions/upload-artifact@v2
+        with:
+          name: coverage
+          path: coverage.info
```
[the remark this answers is not in the document]
Comment (drgorillamd) [reply .github/workflows/coverage_check.yml:73]: artifact-name saves the html report, while I'm uploading the total coverage (as a single int, so easier to compare than having to parse the html) - this should be changed to rather use the branch coverage, but the action maintainer has deserted (I'll fork it one day)

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -64,6 +64,68 @@ jobs:
...
+        run: |
+          forge build --build-info
+
+      - name: Run Echidna
+        uses: crytic/echidna-action@v2
+        with:
+          files: .
+          contract: InvariantGreeter
+          test-mode: assertion
+          crytic-args: --ignore-compile
```
[the remark this answers is not in the document]
Comment (drgorillamd) [reply .github/workflows/tests.yml:101]: you can pass a directory (in files), `contract` is optional

[the remark this answers is not in the document]
Comment (gas1cent) [reply .github/workflows/tests.yml:101]: > you can pass a directory

@drgorillamd let's do that? Will be one less thing for whoever starts with the boilerplate to modify.

[the remark this answers is not in the document]
Comment (drgorillamd) [reply .github/workflows/tests.yml:101]: til: when passing a directory, only the first contract is run, including the import (so here, it tries to run it on IERC20...). For multiple contracts, people are actually using a proxy

**Review by gas1cent — APPROVED** [review]

**Review by 0xOneTony — APPROVED** [review]
🔥

**Review by gas1cent — APPROVED** [review]

File: package.json [hunk 4]
```diff
@@ -22,7 +22,9 @@
     "lint:sol-tests": "solhint -c .solhint.tests.json 'test/**/*.sol'",
     "prepare": "husky install",
     "test": "forge test -vvv",
+    "test:fuzz": "echidna test/invariants/fuzz/Greeter.t.sol --contract GreeterInvariant --corpus-dir test/invariants/fuzz/echidna_coverage/ --test-mode assertion",
```
Comment (0xmoebius) [inline package.json:25]: Is it possible to use wildcards to run against all contracts under `fuzz` ?

Comment (drgorillamd) [reply package.json:25]: check the previous comment from gas @0xmoebius ;)

**Review by 0xmoebius — APPROVED** [review]

**Review by gas1cent — APPROVED** [review]

**Pull request merged**